libsnark/gadgetlib1/gadgets/verifiers/kzg10_batched_verifier_gadget.hpp

```diff
@@ -0,0 +1,261 @@
+/** @file
+ *****************************************************************************
+ * @author     This file is part of libff, developed by Clearmatics Ltd
+ *             (originally developed by SCIPR Lab) and contributors
+ *             (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef LIBSNARK_GADGETLIB1_GADGETS_VERIFIERS_KZG10_BATCHED_VERIFIER_GADGET_HPP_
+#define LIBSNARK_GADGETLIB1_GADGETS_VERIFIERS_KZG10_BATCHED_VERIFIER_GADGET_HPP_
+
+#include "libsnark/gadgetlib1/gadgets/verifiers/kzg10_verifier_gadget.hpp"
+#include "libsnark/polynomial_commitments/kzg10_batched.hpp"
+
+namespace libsnark
+{
+
+// TODO: create an "evaluations variable object"
+
+template<typename ppT> class kzg10_batched_witness_variable
+{
+public:
+    kzg10_witness_variable<ppT> W_1;
+    kzg10_witness_variable<ppT> W_2;
+
+    kzg10_batched_witness_variable(
+        protoboard<libff::Fr<ppT>> &pb, const std::string &annotationn_prefix);
+
+    void generate_r1cs_witness(
+        const typename kzg10_batched_2_point<
+            other_curve<ppT>>::evaluation_witness &eval_witness);
+};
+
+/// Given an array of commitments, and array of evaluations, and some field
+/// element gamma, compute terms of the form:
+///
+///   \sum_{i=1}^{t_1} \gamma^{i-1} (cm_i - [s_i]_1)
+///
+/// The `num_entries` parameter here is intended to reflect the fact that this
+/// must be statically defined, although all internal structures are currently
+/// dynamic. This also allows specialization for the case of 2 entries.
+template<typename ppT, size_t num_entries>
+class kzg10_batched_compute_commit_minus_eval_sum : gadget<libff::Fr<ppT>>
+{
+    static_assert(num_entries > 2, "num_entries must be greater that 2");
+
+public:
+    using Field = libff::Fr<ppT>;
+
+    // These are technically he negative encoded evaluations:
+    //
+    //   encoded_evals[i] = G1_mul(evals[i], -G1::one());
+    std::vector<G1_variable_or_identity<ppT>> encoded_evals;
+    std::vector<G1_mul_by_scalar_gadget<ppT>> compute_encoded_evals;
+
+    // Negative evals are added to commits to compute cm_i - [s_i]_1:
+    //
+    //   commit_minus_encoded_eval[i] = cm_i - [s_i]_1
+    std::vector<G1_variable<ppT>> commit_minus_encoded_eval;
+    std::vector<G1_add_variable_and_variable_or_identity_gadget<ppT>>
+        compute_commit_minus_encoded_eval;
+
+    // TODO: replace this with a calculation of the form:
+    //
+    //   \gamma ( ... ( \gamma (cm_n - s_n) + cm_{n-1} - s_{n-1} ) ...)
+    //
+    // to avoid problems with the incompatibility of inner and outer scalar
+    // fields.
+
+    // gamma_powers does not contain 1 or gamma, so:
+    //
+    //   gamma_powers[0] = gamma^2
+    //   gamma_powers[i>0] = gamma * gamma_powers[i - 1]
+    //
+    // so length = num_entries - 2
+    pb_linear_combination<Field> gamma;
+    pb_variable_array<Field> gamma_powers;
```

Comment:
This is invalid since arithmetic will be performed in the wrapping scalar field - not the inner scalar field.
https://github.com/clearmatics/libsnark/pull/45#issuecomment-1103987863

Reply:
Fixed in latest commit.